core/node/api_server/src/web3/namespaces/en.rs

```diff
@@ -44,18 +37,30 @@ impl EnNamespace {
     }
 
     #[tracing::instrument(skip(self))]
-    pub async fn attestation_status_impl(&self) -> Result<en::AttestationStatus, Web3Error> {
-        Ok(en::AttestationStatus {
-            next_batch_to_attest: to_l1_batch_number(
-                self.state
-                    .acquire_connection()
-                    .await?
-                    .consensus_dal()
-                    .next_batch_to_attest()
-                    .await
-                    .context("next_batch_to_attest()")?,
-            )?,
-        })
+    pub async fn attestation_status_impl(
+        &self,
+    ) -> Result<Option<en::AttestationStatus>, Web3Error> {
+        let status = self
+            .state
+            .acquire_connection()
+            .await?
+            // unwrap is ok, because we start outermost transaction.
+            .transaction_builder()
+            .unwrap()
+            // run readonly transaction to perform consistent reads.
+            .set_readonly()
+            .build()
+            .await
+            .context("TransactionBuilder::build()")?
+            .consensus_dal()
+            .attestation_status()
+            .await?;
+
+        Ok(status.map(|s| {
+            en::AttestationStatus(
+                zksync_protobuf::serde::serialize(&s, serde_json::value::Serializer).unwrap(),
```

Comment:
What is the rationale behind returning a wrapped JSON value here? 

I see this is what the `consensus_genesis_impl` does as well, but it returns `zksync_consensus_roles::validator::messages::consensus::Genesis`, which seems like a public and well-known type, it's obvious what to deserialise it into, as every consensus client should have access to it. Whereas here we are returning a type defined inside the DAL consensus module. Are those supposed to be part of the web interface? Normally I would not expect this to be the case because I don't expect web DTOs to bring in `sqlx` dependencies, which is why I suppose it's returned as JSON. I am guessing that's because the consensus types aren't available where the web types are declared. What are we supposed to deserialise this into? Can't we have a common type declared somewhere accessible to the DAL and the web _client_ at least?

Reply:
There are some people which didn't want me to add consensus crates as a dependency of zksync_types crate. Then someone else came and added zksync_protobuf as dependency of zksync_types anyway. One may argue that zksync_consensus_roles is different (not my opinion) and that's why it shouldn't be a dependency of zksync_types. I didn't want to waste time arguing about that, so just followed a pattern that got accepted.

> Can't we have a common type declared somewhere accessible to the DAL and the web client at least?

I'm open to suggestions where to move it.

Reply:
Thanks for the explanation, I thought it might be something like this. 

I see that `zksync_types` is a dependency for `zksync_dal`, so we could keep the fully fledged `AttestationStatus` where it is now, next to the `SyncBlock`, and return it from the DAL as-is. I'm not sure what that means for the protobuf format which is also defined in the DAL; I don't think it's used for `SyncBlock`, is it? I see `SyncDal::sync_block` converts from `zksync_dal::models::storage_sync::SyncBlock` to `zksync_types::api::en::SyncBlock`, and the former is mapped from a `StorageSyncBlock`, so there is a whole series of DTOs but the point is the DAL returns what the web API serves in this case. Since we query individual fields we could populate it directly. 

I understand why `ConsensusGenesis` return JSON, who would want to repeat all the fields if they don't have to, but I'm not sure we should apply the same pattern just because it's another consensus related thing, while the DTO and the protobuf format that controls its JSON format is inaccessible.

Reply:
I suppose another option is to move the DTO and proto definition over to the consensus crate and use the JSON route in the API. But don't ask me under what package these web-api related types should live.

Reply:
I mean, I treat this endpoint as a transport layer (we could use binary encoding afaic) and I don't really want/need it to be interpreted outside the scope of the consensus.

AFAICT what you are trying to achieve is to use a stronger type in `zksync_types::api::en`. I see a value in that, but the apparent discrimination against adding dependencies to zksync_types (and any non-serde encoding for that matter) is too discouraging for me to bother with that.

Reply:
You are right, I focussed on what would be a nice type for the API to expose, and immediately thought that surely the DAL types are private and won't help to deserialise the data. I forgot that we actually only want to use this API from the node itself, where the DAL _is_ accessible, so I won't have any trouble parsing the JSON regardless of whether this type and its proto resides here, or in the era-consensus repo. Of course keeping it here is more convenient, so I'm happy with it.